compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -0,0 +1,232 @@
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  return replacements;
+}
+
+/// Returns true if the given op is collapsable.
+static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+  // TODO(guray) There is no mechanism to tell the collapsed indexes to
+  // `tensor.expand_shape`. Once we have this support in MLIR, we can enable
+  // dynamic tensor shapes.
+  if (genericOp.hasDynamicShape()) return false;
+
+  // TODO(guray) Currently we can only collapse when result of all the
+  // AffineMaps are dimensions. Possible to collapse cases like
+  // affine_map<d0, d1+d2> with affine_map<d0, d1+d2>, however, this is not
+  // supported in collapsing mechanism in MLIR. Once we have this support,
+  // we can remove this if statement.
+  if (llvm::any_of(genericOp.getIndexingMapsArray(), [](AffineMap map) {
+        return !map.isProjectedPermutation();
+      })) {
+    return false;
+  }
+
+  // IndexOp allows accesing induction variables. Collapsing might cause
+  // performance regression, so we disable it.
+  if (genericOp.hasIndexSemantics()) return false;
+
+  return true;
+}
+
+/// Traverses all the the Ops in DispatchRegionOps and finds linalg.generic Op
+/// without any producers.
+static FailureOr<linalg::GenericOp> findRootGenericOp(
+    DispatchRegionOp regionOp) {
+  auto &ops = regionOp.getBody().front().getOperations();
+  if (ops.size() != 2) return failure();
+
+  linalg::GenericOp genericOp = dyn_cast<linalg::GenericOp>(ops.front());
+  if (!genericOp) return failure();
+
+  return genericOp;
+}
+
+/// Generate a new dispatch.region and workload according with the collapsed
+/// linalg Generic Op
+static LogicalResult generateNewDispatchRegion(
+    IRRewriter &rewriter, DispatchRegionOp regionOp,
+    SmallVector<Value> collapseResults, linalg::GenericOp newGenericOp) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(regionOp->getParentOp());
+
+  auto maybeBuilder =
+      iree_compiler::IREE::Flow::getWorkloadBuilder(rewriter,
+                                                    /*rootOp=*/newGenericOp);
+  if (failed(maybeBuilder)) return failure();
+
+  auto maybeRegionOp =
+      Flow::wrapOpInDispatchRegion(rewriter, newGenericOp, *maybeBuilder);
+  if (failed(maybeRegionOp)) return failure();
+
+  // Replace old regionOp with the result of collapse
+  rewriter.replaceOp(regionOp, collapseResults);
+
+  return success();
+}
+
+/// Traverses DispatchRegionOps to find linalg genericOps that has no
+/// producers and tries to collapse its dimensions.
+static LogicalResult collapseDimensions(IRRewriter &rewriter,
+                                        DispatchRegionOp &regionOp) {
+  // Step 1. Find the root linalg.generic Op with no producer
+  auto maybeGenericOp = findRootGenericOp(regionOp);
```

Comment:
Style nit... Just easier if you just make this
```
std::optional<linalg::GenericOp> genericOp = findRootGenericOp(regionOp);
```

and then use `genericOp->` instead of `genericOp.` eveywhere...